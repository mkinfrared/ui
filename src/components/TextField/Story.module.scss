.Story {
  > div {
    width: 40rem;
  }
}
