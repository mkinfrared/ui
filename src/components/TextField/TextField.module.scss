@import "../../styles/common";

.TextField {
  display: block;
  width: 100%;

  &.hasError {
    .input {
      border-color: var(--error);

      // todo replace with relative colors
      // when browsers start to support it
      box-shadow: 0 0 0.5rem 0 var(--error);
    }
  }

  &.disabled {
    .label {
      color: var(--grey);
      cursor: default;
    }

    .input {
      border-color: var(--grey);
    }

    .input input {
      color: var(--grey);
    }
  }

  &.hasSuffix {
    .input {
      padding-right: 3rem;
    }
  }

  &.hasPrefix {
    .input {
      padding-left: 3rem;
    }
  }

  .label {
    cursor: pointer;
    margin-bottom: 1rem;
    text-transform: capitalize;
  }

  .input {
    border: 1px solid transparent;
    border-radius: 1rem;
    font-family: Roboto, sans-serif;
    font-size: var(--text-size);
    font-style: normal;
    font-weight: 500;
    line-height: 1.8;
    overflow: hidden;
    position: relative;
    width: 100%;

    :where(&) {
      @include dark;

      @include theme("light") {
        @include light;
      }
    }

    input {
      background-color: transparent;
      border: none;
      caret-color: var(--secondary);
      color: inherit;
      font-size: 100%;
      height: 100%;
      line-height: 1;
      padding: 1.5rem 2rem;
      width: 100%;

      &:-webkit-autofill,
      &:-webkit-autofill:active,
      &:-webkit-autofill:hover,
      &:-webkit-autofill:focus {
        background-color: var(--background-color);
        border: none;
        box-shadow: 0 0 0 1000px var(--background-color) inset;
        caret-color: var(--secondary);
        font-size: 100%;
        -webkit-text-fill-color: var(--white);

        // it's not possible to use transparent background color
        // to override chrome autofill styles, so background needs
        // to be set to a certain color, to keep the initial background
        // transparent the transition delay can be used
        transition: background-color 7200s ease-in-out 0s; // 2 hr delay
      }

      &::selection {
        background-color: var(--cyan);
        color: var(--white);
      }

      &:focus-visible {
        outline: none;
      }
    }

    .prefix,
    .suffix {
      aspect-ratio: 1/1;
      height: calc(var(--text-size) * 1.5);
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: calc(var(--text-size) * 1.5);

      svg {
        fill: currentcolor;
        height: 100%;
        width: 100%;
      }
    }

    .prefix {
      left: 1rem;
    }

    .suffix {
      cursor: pointer;
      right: 1rem;
    }
  }

  .error {
    color: var(--error);
    font-size: 1.4rem;
    height: 2.5rem;
    margin-top: 0.5rem;
    overflow: hidden;
    padding: 0;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;
  }
}

label {
}
