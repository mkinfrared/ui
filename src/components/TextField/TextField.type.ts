import { FakeButtonProps } from "components/FakeButton";

type InputAttributes = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "prefix"
>;

export type TextFieldProps = InputAttributes & {
  autoComplete?: string;
  className?: string;
  disabled?: boolean;
  error?: string;
  name?: string;
  inputRef?: React.Ref<HTMLInputElement | null | undefined>;
  label?: string;
  onBlur?: React.FocusEventHandler<HTMLInputElement>;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onSuffixClick?: FakeButtonProps["onClick"];
  prefix?: React.ReactElement<SVGSVGElement>;
  suffix?: React.ReactElement<SVGSVGElement>;
  type?:
    | "email"
    | "file"
    | "hidden"
    | "number"
    | "password"
    | "search"
    | "tel"
    | "text"
    | "url";
};
