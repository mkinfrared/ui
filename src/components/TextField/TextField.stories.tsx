import { action } from "@storybook/addon-actions";
import { Meta, Story } from "@storybook/react";
import React, { useState } from "react";

import { ReactComponent as Doc } from "icons/description.svg";
import { ReactComponent as Eye } from "icons/visibility.svg";

import css from "./Story.module.scss";
import { TextField } from "./TextField";
import { TextFieldProps } from "./TextField.type";

export default {
  title: "UI/TextField",
  component: TextField,
} as Meta;

const Template: Story<TextFieldProps> = (args) => (
  <div className={css.Story}>
    <div>
      <TextField {...args} />
    </div>
  </div>
);

const ControlledTemplate: Story<TextFieldProps> = () => {
  const [value, setValue] = useState("");

  return (
    <div className={css.Story}>
      <div>
        <TextField
          value={value}
          onChange={(event) => setValue(event.target.value)}
        />
      </div>
    </div>
  );
};

const MultiFields: Story<TextFieldProps> = (args) => (
  <div className={css.Story}>
    <div>
      <TextField {...args} />
      <TextField {...args} />
      <TextField {...args} />
      <TextField {...args} />
    </div>
  </div>
);

const Form: Story<TextFieldProps> = () => (
  <div className={css.Story}>
    <div>
      <TextField name="username" />
      <TextField name="password" suffix={<Eye />} />
    </div>
  </div>
);

const Uncontrolled = Template.bind({});
const Controlled = ControlledTemplate.bind({});
const WithError = Template.bind({});
const WithPrefix = Template.bind({});
const WithSuffix = Template.bind({});
const WithPrefixSuffix = Template.bind({});
const MultipleFields = MultiFields.bind({});
const ExampleForm = Form.bind({});

Uncontrolled.args = {
  disabled: false,
  value: "foobar",
  error: "",
  label: "",
};

WithError.args = {
  value: "marklar",
  onChange: action("change"),
  error:
    "Lorem ipsum dolor sit amet, consectetur adipisicing elit. asdf sadfl kjasdf agj asldfkj asdfjlk",
};

WithPrefix.args = {
  value: "marklar",
  onChange: action("change"),
  prefix: <Doc />,
};

WithSuffix.args = {
  value: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
  onChange: action("change"),
  suffix: <Eye />,
  onSuffixClick: action("suffix"),
};

WithPrefixSuffix.args = {
  value: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
  onChange: action("change"),
  suffix: <Eye />,
  prefix: <Doc />,
};

export {
  Uncontrolled,
  Controlled,
  WithError,
  WithPrefix,
  WithSuffix,
  WithPrefixSuffix,
  MultipleFields,
  ExampleForm,
};
