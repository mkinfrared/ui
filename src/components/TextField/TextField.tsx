import { uniqueId } from "lodash-es";
import { useMemo } from "react";

import FakeButton from "components/FakeButton";
import Heading from "components/Heading";
import Text from "components/Text";
import { classNames, mergeRefs } from "utils";

import css from "./TextField.module.scss";
import { TextFieldProps } from "./TextField.type";

const TextField = ({
  autoComplete,
  className,
  disabled = false,
  error,
  inputRef,
  name,
  label,
  type = "text",
  onBlur,
  onChange,
  onSuffixClick,
  prefix,
  suffix,
  value,
}: TextFieldProps) => {
  const classes = [
    css.TextField,
    className,
    !!error && css.hasError,
    !!prefix && css.hasPrefix,
    !!suffix && css.hasSuffix,
    disabled && css.disabled,
  ];

  const errorId = useMemo(() => uniqueId("error"), []);
  const mergedRefs = mergeRefs(inputRef);

  return (
    <label className={classNames(...classes)} data-testid="TextField">
      {label && (
        <Heading variant="h5" className={css.label}>
          {label}
        </Heading>
      )}
      <div className={css.input}>
        {prefix && <FakeButton className={css.prefix}>{prefix}</FakeButton>}
        <input
          aria-invalid={!!error}
          aria-errormessage={errorId}
          autoComplete={autoComplete}
          disabled={disabled}
          name={name}
          onBlur={onBlur}
          onChange={onChange}
          ref={mergedRefs}
          type={type}
          value={value}
        />
        {suffix && (
          <FakeButton className={css.suffix} onClick={onSuffixClick}>
            {suffix}
          </FakeButton>
        )}
      </div>
      <Text aria-live="polite" className={css.error} id={errorId}>
        {error}
      </Text>
    </label>
  );
};

export default TextField;
