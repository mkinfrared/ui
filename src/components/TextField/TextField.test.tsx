import { render } from "@testing-library/react";
import React from "react";

import { TextField } from "./TextField";

describe("<TextField />", () => {
  const Component = <TextField />;

  it("should be defined", () => {
    expect(TextField).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should not have Heading when label is not defined", () => {
    const { queryByTestId } = render(Component);
    const heading = queryByTestId("Heading");

    expect(heading).not.toBeInTheDocument();
  });

  it("should have Heading when label is a truthy string", () => {
    const label = "Marklar";
    const { queryByTestId } = render(<TextField label={label} />);
    const heading = queryByTestId("Heading");

    expect(heading).toBeInTheDocument();

    expect(heading).toHaveTextContent(label);
  });

  it("should have a hasError class name when error is a truthy string", () => {
    const error = "Marklar";
    const { getByTestId } = render(<TextField error={error} />);
    const textField = getByTestId("TextField");

    expect(textField.classList).toContain("hasError");
  });

  it("should have a hasPrefix class name when prefix is defined", () => {
    const testId = "prefix";
    const prefix = <div data-testid={testId}>Marklar</div>;

    const { getByTestId, queryByTestId } = render(
      <TextField prefix={prefix} />,
    );

    const textField = getByTestId("TextField");
    const prefixElement = queryByTestId(testId);

    expect(textField.classList).toContain("hasPrefix");

    expect(prefixElement).toBeInTheDocument();
  });

  it("should have a hasSuffix class name when suffix is defined", () => {
    const testId = "suffix";
    const suffix = <div data-testid={testId}>Marklar</div>;

    const { getByTestId, queryByTestId } = render(
      <TextField suffix={suffix} />,
    );

    const textField = getByTestId("TextField");
    const prefixElement = queryByTestId(testId);

    expect(textField.classList).toContain("hasSuffix");

    expect(prefixElement).toBeInTheDocument();
  });
});
