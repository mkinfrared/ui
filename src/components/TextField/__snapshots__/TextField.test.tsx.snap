// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should match the snapshot 1`] = `
<div>
  <label
    class="TextField"
    data-testid="TextField"
  >
    <div
      class="input"
    >
      <input
        type="text"
        value=""
      />
    </div>
    <p
      class="Text error"
      data-testid="Text"
    />
  </label>
</div>
`;
